plugins/modules/files/archive.py

```diff
@@ -196,447 +194,418 @@
+STATE_INCOMPLETE = 'incomplete'
 
-def to_b(s):
+
+def _to_bytes(s):
     return to_bytes(s, errors='surrogate_or_strict')
 
 
-def to_n(s):
+def _to_native(s):
     return to_native(s, errors='surrogate_or_strict')
 
 
-def to_na(s):
+def _to_native_ascii(s):
     return to_native(s, errors='surrogate_or_strict', encoding='ascii')
 
 
 def expand_paths(paths):
     expanded_path = []
     is_globby = False
     for path in paths:
-        b_path = to_b(path)
+        b_path = _to_bytes(path)
         if b'*' in b_path or b'?' in b_path:
             e_paths = glob.glob(b_path)
             is_globby = True
-
         else:
             e_paths = [b_path]
         expanded_path.extend(e_paths)
     return expanded_path, is_globby
 
 
-def matches_exclusion_patterns(path, exclusion_patterns):
-    return any(fnmatch(path, p) for p in exclusion_patterns)
-
-
-def get_filter(exclusion_patterns, format):
-    def zip_filter(path):
-        return matches_exclusion_patterns(path, exclusion_patterns)
+def is_archive(path):
+    return re.search(br'\.(tar|tar\.(gz|bz2|xz)|tgz|tbz2|zip)$', os.path.basename(path), re.IGNORECASE)
 
-    def tar_filter(tarinfo):
-        return None if matches_exclusion_patterns(tarinfo.name, exclusion_patterns) else tarinfo
 
-    return zip_filter if format == 'zip' or not PY27 else tar_filter
-
-
-def get_archive_contains(format):
-    def archive_contains(archive, name):
-        try:
-            if format == 'zip':
-                archive.getinfo(name)
-            else:
-                archive.getmember(name)
-        except KeyError:
-            return False
+def legacy_filter(path, exclusion_patterns):
+    return matches_exclusion_patterns(path, exclusion_patterns)
 
-        return True
 
-    return archive_contains
+def matches_exclusion_patterns(path, exclusion_patterns):
+    return any(fnmatch(path, p) for p in exclusion_patterns)
 
 
-def get_add_to_archive(format, filter):
-    def add_to_zip_archive(archive_file, path, archive_name):
+@six.add_metaclass(abc.ABCMeta)
+class Archive(object):
+    def __init__(self, module):
+        self.module = module
+
+        self.destination = _to_bytes(module.params['dest']) if module.params['dest'] else None
+        self.exclusion_patterns = module.params['exclusion_patterns'] or []
+        self.format = module.params['format']
+        self.must_archive = module.params['force_archive']
+        self.remove = module.params['remove']
+
+        self.changed = False
+        self.destination_state = STATE_ABSENT
+        self.errors = []
+        self.file = None
+        self.root = b''
+        self.successes = []
+        self.targets = []
+        self.not_found = []
+
+        paths = module.params['path']
+        self.expanded_paths, has_globs = expand_paths(paths)
+        self.expanded_exclude_paths = expand_paths(module.params['exclude_path'])[0]
+
+        self.paths = list(set(self.expanded_paths) - set(self.expanded_exclude_paths))
```

Comment:
Something that escaped me before: do we really need it to be a `list`? Maybe having it as `set` will be enough.

Reply:
Leaving it as a `set` shouldn't break anything at the moment, but I should probably change that `list` to `sorted` so that the ordering is predictable.